Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h

```diff
@@ -0,0 +1,139 @@
+/*=========================================================================
+ *
+ *  Copyright NumFOCUS
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+
+#ifndef itkFFTImageFilterFactory_h
+#define itkFFTImageFilterFactory_h
+
+#include "itkObjectFactoryBase.h"
+#include "itkVersion.h"
+
+namespace itk
+{
+
+/** \class FFTImageFilterTraits
+ *
+ * \brief Helper defining pixel traits for templated FFT image filters
+ *
+ * FFT factory registration must account for whether the input/output
+ * images for an FFT class are real- or complex-valued. Each FFT
+ * image filter must be accompanied by a specialized FFTImageFilterTraits
+ * class to define these mappings for factory registration.
+ *
+ * \ingroup FourierTransform
+ * \ingroup ITKFFT
+ */
+template <template <typename, typename> class TFFTImageFilter>
+struct FFTImageFilterTraits
+{
+  template <typename TUnderlying>
+  using InputPixelType = TUnderlying;
+  template <typename TUnderlying>
+  using OutputPixelType = TUnderlying;
```

Comment:
I agree with the sentiment, but added these because I was getting a compile-time error when input/output types were not defined for the base class even though it should never be directly referenced. I'll check whether this was resolved with a forward declaration.

Reply:
I am hopeful that the compiler errors that we are seeing when the non-specialization does not have the using statements are exactly the ones we hope to see whenever the non-specialization code is being used instead of the overriding specialization code that we hope to be using.  (GCC 10.3.0 on Ubuntu 20.04 does exactly the right thing on my system.)

If there is a compiler out there that doesn't do it the way GCC 10.3.0 is doing it and those `using` statements in the non-specialization code do prove to be necessary for that compiler, I recommend that they instead be set to `void`.  

```c++
  template <typename TUnderlying>
  using InputPixelType = void;
  template <typename TUnderlying>
  using OutputPixelType = void;
```

Hopefully that will cause a compiler error somewhere down the road in the scenario that the non-specialization code is being used instead of the appropriate template specialization code; I would hate to miss this at compile time and have to debug this sort of error at run time.

Reply:
It looks like GCC 9.3.0 agrees, the base template `Traits` class need not define `InputPixelType` and `OutputPixelType` so long as the class is declared before its use in `FFTImageFilterFactory`. I heartily agree that we want to get errors related to `Traits` classes like this at compile rather than run time. I've pushed changes accordingly, will watch CI to verify.